### DISCUSSION TITLE
Preprocessor symbols disappear after project refresh or Eclipse restart

### DISCUSSION THREAD

**Discussion by midiwidi** [eclipse-cdt/cdt#261] [Q&A] [answered]
Before I create an issue, I want to find out if I'm doing something wrong and if this is a CDT bug or not.
I have set up a "C Managed Build" project for an ARM Cortex-M0 using the Arm GNU Toolchain (arm-none-eabi). My project needs a preprocessor symbol (LFS_CONFIG=user_lfs_config.h) to compile without errors.
I set this in the project properties in C/C++ Build -> Settings under GNU Arm Cross C Compiler -> Preprocessor -> Defined Symbols

![image](https://user-images.githubusercontent.com/54610151/215029884-a26f19d7-abbf-423d-baba-83e830a7d3f4.png)

After I "Apply and Close", my project builds without errors.
But if I close Eclipse and open it again or if I refresh the project (using F5) then the build fails. If I look into the settings again I can see that the symbol has  disappeared.
Interestingly, there are more symbols which were automatically generated (likely by the CMSIS RTE) which don't disappear.

![image](https://user-images.githubusercontent.com/54610151/215032623-cf95d83d-6c00-4575-b448-d3043a911946.png)

The same thing happens if I define the symbol in C/C++ General -> Paths and Symbols under # Symbols.

Does anybody else have a similar problem and know how to solve this?

Eclipse and all plugins should be up-to-date. The CDT core version is
"Eclipse CDT	C/C++ Development Tools Core	8.0.0.202211292120	org.eclipse.cdt.core"

**Comment by jonahgraham** [comment]
This sounds like a bug, but probably needs to be filed with https://github.com/eclipse-embed-cdt/eclipse-plugins/issues? I assume you have those plug-ins installed too? Do you have any other third-party plug-ins installed?

**Comment by midiwidi** [comment, marked as answer]
Yes, I have Eclipse Embedded CDT installed. I also have other third-party plug-ins installed (see the list of them at the end of the post).

I did some more testing and found out that it depends on the project type if this bug shows up or not. I used this simple "hello world" program
```C
#include <stdio.h>

#ifdef TEST_DEFINE
int main(void)
{
	printf("Hello Arm World!" "\n");
	return 0;
}
#endif
```
to create 4 different projects (test1 ... test4) with different project types and toolchains. All of them were CDT managed C builds.

![image](https://user-images.githubusercontent.com/54610151/215632597-42e17b90-801c-4971-8653-95c1fa42277b.png)

The problem with the preprocessor symbol being dropped after refresh or Eclipse restart only appeared in the project test4 (CMSIS C/C++ Project). This possibly means that the problem is caused by the CMSIS-Pack Eclipse plug-in.


**Edit:**
Looking at the [CMSIS-Pack Eclipse plug-in issues](https://github.com/ARM-software/cmsis-pack-eclipse/issues/88) finally pointed me to the solution for this. The user defined symbols have to be moved above the CMSIS created _RTE_ symbol. All symbols below it are deleted/refresh each time Eclipse starts or the project is refreshed.


My installed Eclipse plugins
|Provider|Name|Version|ID|
|--------|----|-------|--|
|ARM Limited|CMSIS-Pack Eclipse plug-ins|2.7.0.202206081009|com.arm.cmsis.pack.feature.feature.group|
|ARM Limited|CMSIS-Pack Eclipse plug-ins (source code)|2.7.0.202206081009|com.arm.cmsis.pack.feature.source.feature.group|
|ARM Limited|CMSIS-Zone Eclipse plug-ins|1.2.1.202206081009|com.arm.cmsis.zone.feature.feature.group|
|Eclipse CDT|C/C++ Autotools support|11.0.0.202211091556|org.eclipse.cdt.autotools.feature.group|
|Eclipse CDT|C/C++ CMake Build Support - Preview|11.0.0.202211080413|org.eclipse.cdt.cmake.feature.group|
|Eclipse CDT|C/C++ Development Tools|11.0.0.202211300214|org.eclipse.cdt.feature.group|
|Eclipse CDT|C/C++ Docker Container Launch Support|11.0.0.202211290115|org.eclipse.cdt.docker.launcher.feature.group|
|Eclipse CDT|C/C++ GCC Cross Compiler Support|11.0.0.202211062329|org.eclipse.cdt.build.crossgcc.feature.group|
|Eclipse CDT|C/C++ GDB Hardware Debugging|11.0.0.202211221643|org.eclipse.cdt.debug.gdbjtag.feature.group|
|Eclipse CDT|C/C++ Launch over Serial|11.0.0.202211062329|org.eclipse.cdt.launch.serial.feature.feature.group|
|Eclipse CDT|C/C++ Memory View Enhancements|11.0.0.202211062329|org.eclipse.cdt.debug.ui.memory.feature.group|
|Eclipse CDT|C/C++ Meson Build Support - Preview|11.0.0.202211091556|org.eclipse.cdt.meson.feature.group|
|Eclipse CDT|C/C++ Remote Launch|11.0.0.202211062329|org.eclipse.cdt.launch.remote.feature.group|
|Eclipse CDT|C/C++ Standalone Debugger|11.0.0.202211211546|org.eclipse.cdt.debug.standalone.feature.group|
|Eclipse CDT|C/C++ Unit Testing Support|11.0.0.202211062329|org.eclipse.cdt.testsrunner.feature.feature.group|
|Eclipse CDT|LaunchBar|11.0.0.202211062329|org.eclipse.launchbar.feature.group|
|Eclipse CDT|Remote Launch Targets|11.0.0.202211062329|org.eclipse.launchbar.remote.feature.group|
|Eclipse CDT|TM Terminal|11.0.0.202210031642|org.eclipse.tm.terminal.feature.feature.group|
|Eclipse CDT|TM Terminal Serial Connector Extensions|11.0.0.202211062329|org.eclipse.tm.terminal.connector.cdtserial.feature.feature.group|
|Eclipse EGit|Git integration for Eclipse|6.4.0.202211300538-r|org.eclipse.egit.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ ADuCM360 Project Template|6.3.1.202210101738|org.eclipse.embedcdt.templates.ad.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ Arm Cross Compiler|6.3.1.202210101738|org.eclipse.embedcdt.managedbuild.cross.arm.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ CMSIS Packs (Experimental)|6.3.1.202210101738|org.eclipse.embedcdt.packs.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ CMSIS Packs (Experimental) Developer Resources|6.3.1.202210101738|org.eclipse.embedcdt.packs.source.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ Core|6.3.1.202210101738|org.eclipse.embedcdt.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ Documentation (Placeholder)|6.3.1.202210101738|org.eclipse.embedcdt.doc.user.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ Freescale Project Templates|6.3.1.202210101738|org.eclipse.embedcdt.templates.freescale.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ GDB JTAG Debugging|6.3.1.202210101738|org.eclipse.embedcdt.debug.gdbjtag.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ Generic Arm Cortex-M Project Template|6.3.1.202210101738|org.eclipse.embedcdt.templates.cortexm.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ J-Link Debugging|6.3.1.202210101738|org.eclipse.embedcdt.debug.gdbjtag.jlink.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ OpenOCD Debugging|6.3.1.202210101738|org.eclipse.embedcdt.debug.gdbjtag.openocd.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ PyOCD Debugging|6.3.1.202210101738|org.eclipse.embedcdt.debug.gdbjtag.pyocd.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ QEMU Debugging|6.3.1.202210101738|org.eclipse.embedcdt.debug.gdbjtag.qemu.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ RISC-V Cross Compiler|6.3.1.202210101738|org.eclipse.embedcdt.managedbuild.cross.riscv.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ SiFive RISC-V Project Template|6.3.1.202210101738|org.eclipse.embedcdt.templates.sifive.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ STM32Fx Project Templates|6.3.1.202210101738|org.eclipse.embedcdt.templates.stm.feature.group|
|Eclipse Embedded CDT|Embedded C/C++ xPack Arm & RISC-V Project Templates|6.3.1.202210101738|org.eclipse.embedcdt.templates.xpack.feature.group|
|Eclipse Linux Tools|C/C++ Library API Documentation Hover Help|8.9.0.202211291906|org.eclipse.linuxtools.cdt.libhover.feature.feature.group|
|Eclipse Linux Tools|ChangeLog Management Tools for C/C++|8.9.0.202211291906|org.eclipse.linuxtools.changelog.c.feature.group|
|Eclipse Linux Tools|GCov Integration|8.9.0.202211291906|org.eclipse.linuxtools.gcov.feature.group|
|Eclipse Linux Tools|GProf Integration|8.9.0.202211291906|org.eclipse.linuxtools.gprof.feature.feature.group|
|Eclipse Linux Tools|Library Hover help for devhelp documentation|8.9.0.202211291906|org.eclipse.linuxtools.cdt.libhover.devhelp.feature.feature.group|
|Eclipse Linux Tools|RPM Tools|8.9.0.202211291906|org.eclipse.linuxtools.rpm.feature.group|
|Eclipse Linux Tools|Valgrind Tools Integration|8.9.0.202211291906|org.eclipse.linuxtools.valgrind.feature.group|
|Eclipse Marketplace Client|Marketplace Client|1.10.1.v20221110-1841|org.eclipse.epp.mpc.feature.group|
|Eclipse Mylyn|Mylyn Context Connector: C/C++ Development|5.21.2.v20200828-1617|org.eclipse.cdt.mylyn.feature.group|
|Eclipse Mylyn|Mylyn Context Connector: Eclipse IDE|3.25.2.v20200828-1617|org.eclipse.mylyn.ide_feature.feature.group|
|Eclipse Mylyn|Mylyn Context Connector: Team Support|3.25.2.v20200828-1617|org.eclipse.mylyn.team_feature.feature.group|
|Eclipse Mylyn|Mylyn Task List|3.25.2.v20200814-0512|org.eclipse.mylyn_feature.feature.group|
|Eclipse Mylyn|Mylyn Task-Focused Interface|3.25.2.v20200828-1617|org.eclipse.mylyn.context_feature.feature.group|
|Eclipse Mylyn|Mylyn Tasks Connector: Bugzilla|3.25.2.v20200814-0512|org.eclipse.mylyn.bugzilla_feature.feature.group|
|Eclipse Mylyn|Mylyn WikiText|3.0.42.202201072301|org.eclipse.mylyn.wikitext_feature.feature.group|
|Eclipse Nebula|Nebula Widgets|2.7.2.202301171704|org.eclipse.nebula.feature.feature.group|
|Eclipse Nebula|Nebula Widgets Examples|1.1.0.202301171704|org.eclipse.nebula.examples.release.feature.feature.group|
|Eclipse Packaging Project|Eclipse IDE for C/C++ Developers|4.26.0.20221201-1200|epp.package.cpp|
|Eclipse Trace Compass|GDB Tracepoint Analysis|8.2.0.202211302020|org.eclipse.tracecompass.gdbtrace.feature.group|
|Eclipse Trace Compass|LTTng Kernel Analysis|8.2.0.202211302020|org.eclipse.tracecompass.lttng2.kernel.feature.group|
|Eclipse Trace Compass|LTTng Tracer Control|8.2.0.202211302020|org.eclipse.tracecompass.lttng2.control.feature.group|
|Eclipse Trace Compass|LTTng Userspace Analysis|8.2.0.202211302020|org.eclipse.tracecompass.lttng2.ust.feature.group|
|EmbSysRegView Project|Embedded Systems Register View (SFR)|0.2.6|org.eclipse.cdt.embsysregview_feature.feature.group|
|EmbSysRegView Project|EmbSysRegView Data|0.2.6.r191|org.eclipse.cdt.embsysregview.data_feature.feature.group|
|Fabio Zadrozny|PyDev for Eclipse|10.1.3.202301231743|org.python.pydev.feature.feature.group|
|Fabio Zadrozny|Pydev Mylyn Integration|0.6.0|org.python.pydev.mylyn.feature.feature.group|
|null|zamiaCAD|0.11.5|zamiaCAD.feature.group|
|NXP Semiconductors|CodeWarrior Debugger Flash Programmer SDK|14.0.0.202007011858|com.freescale.core.debugger.fpsdk.feature.feature.group|
|NXP Semiconductors, Inc.|FreeRTOS Task Aware Debugger for GDB|1.0.2.201704260904|com.nxp.freertos.gdb.tad.feature.feature.group|
|Sloeber.io|Sloeber, the Eclipse Arduino IDE (plugin version)|4.4.1.202209072349|io.sloeber.feature.feature.group|

**Reply by jonahgraham** [reply]
Great. I am glad that you found the solution and shared it with everyone for the future!
